### PR TITLE
feat(internal): perform changelog retrieval before branch creation

### LINKED ISSUE #5774: Refactor: move changelog retrieval to post-lookup / pre-commit
Currently, changelog retrieval is done "just in time" right before the PR is created. Let's move the existing logic so that it is done after lookups are complete and we know the list of upgrades, but before we create the branch / make a commit.

Doing it this way would mean that release notes could be used/exposed in the commit body if desired.

Todo:

- [ ] work out what structure/fields it can be stored in, to be passed through to the PR creation function
- [ ] make sure this doesn't result in more requests than before
- [ ] make sure we don't accidentally log the entire release notes anywhere now that it's embedded into the `config` that's passed

### PR DESCRIPTION
**Pull request by YuraBeznos**
Refactor: move changelog retrieval to post-lookup / pre-commit (second try with fix)

Closes #5774

### CODE REVIEW

**Comment by rarkins** [comment]
@YuraBeznos can you add us with write access to your fork so that we are able to do things like "Update branch"?

**Comment by YuraBeznos** [comment]
@rarkins I've added you and @viceice with Write access, is it OK?

**Comment by YuraBeznos** [comment]
So right now it is possible to have changelog in commit message something like this:
```
"commitBody": "{{#logJSON.versions}}{{version}} {{date}}\n{{releaseNotes.body}}{{/logJSON.versions}}"
```

**Comment by YuraBeznos** [comment]
@rarkins , @viceice please have a look. Tests are fixed. Feature works as expected.

File: lib/workers/repository/process/extract-update.ts [hunk 1]
```diff
@@ -9,16 +9,17 @@ import { PackageFile } from '../../../manager/common';
...
+export type ExtractResult = {
   branches: BranchConfig[];
   branchList: string[];
-  packageFiles?: Record<string, PackageFile[]>;
+  packageFiles: Record<string, PackageFile[]>;
+};
+
+export type UpdateResult = {
+  res: WriteUpdateResult | undefined;
 };
```
Comment (viceice) [inline lib/workers/repository/process/extract-update.ts:20]: I think we should directly return  `WriteUpdateResult | undefined` and remove this wrapper completely

File: lib/workers/repository/process/extract-update.ts [hunk 2]
```diff
@@ -30,10 +31,20 @@ export async function extractAndUpdate(
...
   sortBranches(branches);
+  return { branches, branchList, packageFiles };
+}
+
+export async function update(
+  config: RenovateConfig,
+  branches: BranchConfig[],
+  branchList: string[],
+  packageFiles: Record<string, PackageFile[]>
+): Promise<UpdateResult> {
```
Comment (viceice) [inline lib/workers/repository/process/extract-update.ts:42]:
```suggestion
): Promise<WriteUpdateResult | undefined> {
```

File: lib/workers/repository/process/extract-update.ts [hunk 3]
```diff
@@ -30,10 +31,20 @@ export async function extractAndUpdate(
...
   let res: WriteUpdateResult | undefined;
   // istanbul ignore else
   if (config.repoIsOnboarded) {
     res = await writeUpdates(config, packageFiles, branches);
   }
-  return { res, branches, branchList, packageFiles };
+
+  return { res };
```
Comment (viceice) [inline lib/workers/repository/process/extract-update.ts:49]:
```suggestion
  return res;
```

File: lib/workers/repository/process/extract-update.ts (hunk 1 above)
Comment (viceice) [reply lib/workers/repository/process/extract-update.ts:20]:
```suggestion
```

File: lib/workers/repository/process/index.ts [hunk 4]
```diff
@@ -56,13 +56,24 @@ export async function processRepo(
...
-  return extractAndUpdate(config);
+  return extract(config);
+}
+
+export async function updateRepo(
+  config: RenovateConfig,
+  branches: BranchConfig[],
+  branchList: string[],
+  packageFiles?: Record<string, PackageFile[]>
+): Promise<UpdateResult> {
```
Comment (viceice) [inline lib/workers/repository/process/index.ts:75]:
```suggestion
): Promise<WriteUpdateResult | undefined> {
```

File: lib/workers/repository/index.ts [hunk 5]
```diff
@@ -31,10 +31,14 @@ export async function renovateRepository(
     await fs.ensureDir(config.localDir);
     logger.debug('Using localDir: ' + config.localDir);
     config = await initRepo(config);
-    const { res, branches, branchList, packageFiles } = await processRepo(
-      config
-    );
+    const { branches, branchList, packageFiles } = await processRepo(config);
     await ensureOnboardingPr(config, packageFiles, branches);
+    const { res } = await updateRepo(
```
Comment (viceice) [inline lib/workers/repository/index.ts:36]:
```suggestion
    const res = await updateRepo(
```

**Review by viceice — CHANGES_REQUESTED** [review]
Maybe needs test fixes after changes

**Comment by YuraBeznos** [comment]
@viceice thanks for suggestions. Everything fixed. Tests are fine.

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@YuraBeznos Have you done some real tests?

**Comment by YuraBeznos** [comment]
@viceice 
Tested: https://github.com/YuraBeznos/sendgrid-nodejs/pulls
Looks good.

**Pull request merged**